Support custom types (eg type customString string)

DarkDrim opened the pull request:
Proposal: In each ScanRow() check if destination type is implementing sql.Scanner
This PR will add support of custom types, eg:
```
type myStr string
type myInt int32
type myFloat float64
```
One requirement: implement sql.Scanner interface. 

Resolves https://github.com/ClickHouse/clickhouse-go/issues/845

---

Also I found unreachable code:
[lib/column/array](https://github.com/ClickHouse/clickhouse-go/blob/main/lib/column/array.go#L243-L246)
Contains unreachable code

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/clickhouse-go?pullRequest=850) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/clickhouse-go?pullRequest=850) before we can accept your contribution.<br/><hr/>**Denys Adamenko** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/clickhouse-go?pullRequest=850) it.</sub>

jkaflik commented:
Hi @DarkDrim 
Thanks for your contribution. Can you provide a set of tests? It's also helps to understand the use case better.

DarkDrim commented:
> Hi @DarkDrim Thanks for your contribution. Can you provide a set of tests? It's also helps to understand the use case better.

sure, do I need to remove unreachable code in array.go?

jkaflik commented:
@DarkDrim that's not necessary. Thanks for letting us know

DarkDrim commented on tests/string_test.go:
```diff
@@ -66,6 +66,61 @@ func TestSimpleString(t *testing.T) {
+	batch, err := conn.PrepareBatch(ctx, "INSERT INTO test_string")
+	require.NoError(t, err)
+
+	type data struct {
+		Col1 string    `ch:"Col1"`
+		Col2 customStr `ch:"Col2"`
+	}
+	require.NoError(t, batch.AppendStruct(&data{
+		Col1: "A",
+		Col2: "B",
+	}))
+	require.NoError(t, batch.Send())
+
+	var dest data
+	require.NoError(t, conn.QueryRow(ctx, "SELECT * FROM test_string").ScanStruct(&dest))
+	assert.Equal(t, "A", dest.Col1)
+	assert.Equal(t, customStr("B"), dest.Col2)
```
Added simple example: now it's possible to use custom string types (as in this example - customStr)
It could be useful when you have some struct with custom string/or other types
To not add some kind of converter that will do cast types.

But it should works with array-type too.
For example when you have:
Array(String)

```
type CustomStrArr []string
```

DarkDrim replied:
> yes. would you mind adding a custom array string test case?

Sure, added

DarkDrim commented on tests/enum_test.go:
```diff
@@ -57,6 +57,47 @@ func TestSimpleEnum(t *testing.T) {
+	ctx := context.Background()
+	require.NoError(t, err)
+	const ddl = `
+			CREATE TABLE test_enum (
+				  Col1 Enum ('hello'   = 1,  'world' = 2),
+				  Col2 Enum8 ('hello'   = 1,  'world' = 2),
+				  Col3 Enum16 ('hello'   = 1,  'world' = 2)
+			) Engine MergeTree() ORDER BY tuple()
+		`
+	defer func() {
+		conn.Exec(ctx, "DROP TABLE IF EXISTS test_enum")
+	}()
+	require.NoError(t, conn.Exec(ctx, ddl))
+	batch, err := conn.PrepareBatch(ctx, "INSERT INTO test_enum")
+	require.NoError(t, err)
+	var (
+		col1Data = customStr("hello")
+		col2Data = customStr("world")
+		col3Data = customStr("hello")
+	)
+	require.NoError(t, batch.Append(
+		col1Data,
+		col2Data,
+		col3Data,
+	))
+	require.NoError(t, batch.Send())
+	var (
+		col1 customStr
+		col2 customStr
+		col3 customStr
+	)
+	require.NoError(t, conn.QueryRow(ctx, "SELECT * FROM test_enum").Scan(&col1, &col2, &col3))
```
Example #2
Enum can be used with custom string types

DarkDrim commented:
I want to add one more example with arrays
Will try to do it soon

DarkDrim commented on lib/column/enum16.go:
```diff
@@ -210,10 +214,14 @@ func (col *Enum16) AppendRow(elem interface{}) error {
-			Op:   "AppendRow",
-			To:   "Enum16",
-			From: fmt.Sprintf("%T", elem),
+		if s, ok := elem.(fmt.Stringer); ok {
+			return col.AppendRow(s.String())
+		} else {
```
Additional check if element implement Stringer - we can use it too

DarkDrim commented on lib/column/enum8.go:
```diff
@@ -216,10 +220,14 @@ func (col *Enum8) AppendRow(elem interface{}) error {
-			Op:   "AppendRow",
-			To:   "Enum8",
-			From: fmt.Sprintf("%T", elem),
+		if s, ok := elem.(fmt.Stringer); ok {
+			return col.AppendRow(s.String())
+		} else {
```
Additional check if element implement Stringer - we can use it too

DarkDrim commented on lib/column/fixed_string.go:
```diff
@@ -149,10 +153,14 @@ func (col *FixedString) AppendRow(v interface{}) (err error) {
-			Op:   "AppendRow",
-			To:   "FixedString",
-			From: fmt.Sprintf("%T", v),
+		if s, ok := v.(fmt.Stringer); ok {
+			return col.AppendRow(s.String())
+		} else {
```
Additional check if element implement Stringer - we can use it too

DarkDrim commented on lib/column/tuple.go:
```diff
@@ -196,8 +196,10 @@ func setJSONFieldValue(field reflect.Value, value reflect.Value) error {
 	// check if our target is a string
 	if field.Kind() == reflect.String {
-		field.Set(reflect.ValueOf(fmt.Sprint(value.Interface())))
-		return nil
+		if v := reflect.ValueOf(fmt.Sprint(value.Interface())); v.Type().AssignableTo(field.Type()) {
+			field.Set(v)
+			return nil
+		}
```
Super important:
see test `TestCustomArray`
in case we have custom type instead of string -> field.Set(v) , where field is a custom string type and v is string type - reflect package will panic
So to fix this possible issue firstly we're checking if value can be assignable to field (if not - just go to next condition and trying to convert val to field).

DarkDrim replied:
Will prepare go/play in a moment to show this issue

DarkDrim replied:
Additional explanation: https://go.dev/play/p/UrtEjsnZEG0

DarkDrim commented on tests/array_test.go:
```diff
@@ -66,6 +66,46 @@ func TestSimpleArray(t *testing.T) {
 	require.NoError(t, rows.Err())
 }
 
+type customArr []customStr
+
+func TestCustomArray(t *testing.T) {
+	conn, err := GetNativeConnection(nil, nil, &clickhouse.Compression{
```
example of usage # 3 - for column Array(Enum) we can use custom string type
It could be useful

DarkDrim commented on lib/column/array.go:
```diff
@@ -240,10 +240,6 @@ func (col *Array) scan(sliceType reflect.Type, row int) (reflect.Value, error) {
 		}
 		return subSlice, nil
 	}
-	return reflect.Value{}, &Error{
-		ColumnType: fmt.Sprint(sliceType.Kind()),
-		Err:        fmt.Errorf("column %s - needs a slice or interface{}", col.Name()),
-	}
```
Removed unreachable code

DarkDrim commented on lib/column/array.go:
```diff
@@ -352,7 +348,6 @@ func (col *Array) scanSliceOfObjects(sliceType reflect.Type, row int) (reflect.V
 				Err:        fmt.Errorf("column %s - needs a slice of objects or an interface{}", col.Name()),
 			}
 		}
-		return reflect.Value{}, nil
```
Removed unreachable code

jkaflik reviewed (approved):
I like the changes. If possible, I'd add one or two test cases for generated columns.

DarkDrim commented:
> I like the changes. If possible, I'd add one or two test cases for generated columns.

Added support for base types
It took a little fiddling, but everything works well
Due to the fact that, for example, int to int64 conversion should fail - I had to add one more extra check

Also added a few tests, will comment them in a few moments

DarkDrim commented on lib/column/codegen/column.tpl:
```diff
@@ -378,10 +381,14 @@ func (col *{{ .ChType }}) AppendRow(v interface{}) error {
-			Op:   "AppendRow",
-			To:   "{{ .ChType }}",
-			From: fmt.Sprintf("%T", v),
+		if rv := reflect.ValueOf(v); rv.Kind() == col.ScanType().Kind() && rv.CanConvert(col.ScanType()) {
+			col.col.Append(rv.Convert(col.ScanType()).Interface().({{ .GoType }}))
+		} else {
+			return &ColumnConverterError{
+				Op:   "AppendRow",
+				To:   "{{ .ChType }}",
+				From: fmt.Sprintf("%T", v),
+			}
```
Supporting base types:
I had to add additional check
```
rv.Kind() == col.ScanType().Kind()
```

because you have one test (SimpleInt) that should fail (int should not be converted into int64)

so how it works:

when we have type int64 and custom int64 type:
```
type customInt64 int64
```

var c customInt64
reflect.TypeOf(c).Kind() == reflect.Int64

that's why it works well

from documentation:

```
The second property is that the Kind of a reflection object describes the underlying type, not the static type. If a reflection object contains a value of a user-defined integer type, as in

type MyInt int
var x MyInt = 7
v := reflect.ValueOf(x)
the Kind of v is still reflect.Int, even though the static type of x is MyInt, not int. In other words, the Kind cannot discriminate an int from a MyInt even though the Type can.
```

https://go.dev/blog/laws-of-reflection

gingerwizard replied:
nice

DarkDrim commented on tests/array_test.go:
```diff
@@ -66,6 +66,50 @@ func TestSimpleArray(t *testing.T) {
+	require.NoError(t, err)
+	const ddl = `
+		CREATE TABLE test_array (
+			  Col1 Array(Enum ('hello'   = 1,  'world' = 2)),
+			  Col2 Array(String)
+		) Engine MergeTree() ORDER BY tuple()
```
Updated test: added Array(Enum) and Array(String)
to show simple []customStrArray not only enum

DarkDrim commented on tests/base_types_test.go:
```diff
@@ -40,6 +50,7 @@ func TestUInt8(t *testing.T) {
 				, Col2 Nullable(UInt8)
 				, Col3 Array(UInt8)
 				, Col4 Array(Nullable(UInt8))
+				, Col5 UInt8
```
Added new test with custom uint8 type

DarkDrim commented on tests/float_test.go:
```diff
@@ -60,6 +60,59 @@ func TestSimpleFloat(t *testing.T) {
+		return
+	}
+	const ddl = `
+		CREATE TABLE test_float (
+			  Col1 Float32,
+			  Col2 Float64
+		) Engine MergeTree() ORDER BY tuple()
```
Added new test with custom float32 and float64

DarkDrim commented on examples/clickhouse_api/custom_types.go:
```diff
@@ -0,0 +1,88 @@
+	return string(s)
+}
+
+func CustomTypes() error {
+	conn, err := GetNativeConnection(nil, nil, nil)
+	if err != nil {
+		return err
+	}
+	ctx := context.Background()
+	defer func() {
+		conn.Exec(context.Background(), "DROP TABLE example")
+	}()
+	if err := conn.Exec(ctx, `DROP TABLE IF EXISTS example`); err != nil {
+		return err
+	}
+	err = conn.Exec(ctx, `
+		CREATE TABLE IF NOT EXISTS example (
+			  Col1 String,
+			  Col2 Enum ('hello'   = 1,  'world' = 2),
+		) Engine = Memory
+	`)
```
Besides tests I added example how to works with custom types (in this example String and Enum)

jkaflik replied:
Thank you. This is something I wanted to ask for. Will review this today. :)

DarkDrim reviewed (commented):
With strings, it turned out much easier, because we can simply check for the fmt.Stringer interface and, if it is defined, use it
For other types, this did not work out, so reflection is used, but this is not the only place with reflection in this CH-client))

so should not be a problem

---

Plus, reflection is used at the very end, when the client would simply return an error
Therefore, there is no problem with performance degradation of base types

jkaflik commented:
@DarkDrim can you have a look at the failed test? don't care about `integration-tests-cloud`. it fails as run on your fork (you don't have access to secrets)

DarkDrim commented:
> @DarkDrim can you have a look at the failed test? don't care about `integration-tests-cloud`. it fails as run on your fork (you don't have access to secrets)

Can you re-run CI please

jkaflik commented:
This solves #753

arianitu commented:
@jkaflik I'm currently still getting this on an Insert and I am using version 


```
{"level":"fatal","error":"clickhouse [AppendRow]: rating clickhouse [AppendRow]: converting model.Rating to String is unsupported"
```

type Rating string

type Organization struct {
    Rating Rating
}

I am also using https://github.com/go-gorm/clickhouse but I think this error comes from this driver?

Am I doing something wrong?

arianitu commented:
I've fixed this in a weird way, is this really necessary to make custom types work, what's going on here? 

```
type Rating string

func (p Rating) String() string {
	return string(p)
}
```


since I read the `func (col *String) AppendRow(v any) error {` code and it has this line:

```
default:
		if valuer, ok := v.(driver.Valuer); ok {
			val, err := valuer.Value()
			if err != nil {
				return &ColumnConverterError{
					Op:   "AppendRow",
					To:   "String",
					From: fmt.Sprintf("%T", v),
					Hint: "could not get driver.Valuer value",
				}
			}
			return col.AppendRow(val)
		}

		if s, ok := v.(fmt.Stringer); ok {
			return col.AppendRow(s.String())
		}

		return &ColumnConverterError{
			Op:   "AppendRow",
			To:   "String",
			From: fmt.Sprintf("%T", v),
		}
	}
```

SpencerTorres commented:
Hello!

It looks like this PR was mainly focused on the scanning, not appending. Your fix appears to be the correct approach based on the presence of this block:
```go
if s, ok := v.(fmt.Stringer); ok {
    return col.AppendRow(s.String())
}
```

This has to do with Go's [type identity](https://go.dev/ref/spec#Type_identity). In the internal `Append` functions the type is switched on the interface as in the example below. In this case the type is not considered a `string`. Since you can't add your own type to this `switch`, the next best option is to use the `fmt.Stringer` interface.

Example ([go playground](https://go.dev/play/p/_xJm843K7l7)):
```go

type Rating string

type Org struct {
	Rating Rating
}

func main() {
	org := Org{
		Rating: "rating!",
	}

	append(org.Rating)
}

func append(v any) {
	switch v.(type) {
	case string:
		fmt.Println("custom type is string")
	case Rating:
		fmt.Println("custom type is Rating")
	default:
		fmt.Println("custom type is not string")
	}
}
```

prints
```
custom type is Rating
```
